modules/pam_setquota/pam_setquota.c

```diff
@@ -0,0 +1,238 @@
+    } else if (strncmp(*argv, "itime=", 6) == 0) {
+      p->dqb_itime = strtol(*argv + 6, NULL, 10);
+      p->dqb_valid |= QIF_ITIME;
+    }
+  }
+
+  return !(bhard ^ bsoft) && !(ihard ^ isoft);
+}
+
+static void _pam_parse_params(int argc, const char **argv,
+                              struct pam_params *p) {
+  for (; argc-- > 0; ++argv) {
+    if (strncmp(*argv, "startuid=", 9) == 0)
+      p->start_uid = strtol(*argv + 9, NULL, 10);
+    else if (strncmp(*argv, "enduid=", 7) == 0)
+      p->end_uid = strtol(*argv + 7, NULL, 10);
+    else if (strncmp(*argv, "fs=", 3) == 0)
+      p->fs = (char *)*argv + 3;
+    else if (strncmp(*argv, "overwrite=", 10) == 0)
+      p->overwrite = strtol(*argv + 10, NULL, 10);
+    else if (strncmp(*argv, "debug=", 6) == 0)
+      p->debug = strtol(*argv + 6, NULL, 10);
+  }
+}
+
+PAM_EXTERN int pam_sm_open_session(pam_handle_t *pamh, int flags, int argc,
+                                   const char **argv) {
+  int retval;
+	char *val;
+  const void *user;
+  const struct passwd *pwd;
+  struct pam_params param = { .start_uid = 1000, .end_uid = 0, .fs = NULL };
+  struct if_dqblk ndqblk;
+  FILE *fd;
+  char mntdevice[BUFSIZ], mntpoint[BUFSIZ];
+  const struct mntent *mnt;
+  const char *service;
+
+  if (pam_get_item(pamh, PAM_SERVICE, (const void **)&service) != PAM_SUCCESS)
+    service = "";
+
+  /* Get UID_MIN for default start_uid from login.defs */
+  val = pam_modutil_search_key(pamh, LOGIN_DEFS, "UID_MIN");
+
+  /* Should UID_MIN be undefined, use 1000 as safe starting UID
+   * to avoid setting a quota for root and system users if startuid= parameter
+   * is absent.
+   */
+  if (val) {
+    param.start_uid = strtol(val, NULL, 10);
+    free(val);
+  }
+  else
+    param.start_uid = 1000;
+
+  /* Parse values */
+  _pam_parse_params(argc, argv, &param);
+
+  /* Determine the user name so we can get the home directory */
+  retval = pam_get_item(pamh, PAM_USER, &user);
+  if (retval != PAM_SUCCESS || user == NULL || *(const char *)user == '\0') {
+    pam_syslog(pamh, LOG_NOTICE, "user unknown");
+    return PAM_USER_UNKNOWN;
+  }
+
+  /* Get the password entry */
+  pwd = getpwnam(user);
+  if (pwd == NULL) {
+    return PAM_CRED_INSUFFICIENT;
+  }
+
+  /* Check if we should not set quotas for user */
+  if ((pwd->pw_uid < param.start_uid) ||
+      ((param.end_uid >= param.start_uid) && (param.start_uid != 0) &&
+       (pwd->pw_uid > param.end_uid)))
+    return PAM_SUCCESS;
+
+  /* Remove the unnecessary '/' from the end of fs parameter */
+  if (param.fs != NULL) {
+    size_t len = strlen(param.fs);
+    if (len > 1)
+      if (param.fs[len - 1] == '/')
+        param.fs[len - 1] = '\0';
+  }
+
+  /* Find out what device the filesystem is hosted on */
+  if ((fd = setmntent("/etc/mtab", "r")) == NULL) {
+    pam_syslog(pamh, LOG_ERR, "Unable to open /etc/mtab");
+    return PAM_PERM_DENIED;
+  }
+
+  *mntpoint = *mntdevice = '\0';
+  size_t match_size = 0;
```

Comment:
Here is another variable declaration in the middle of a block.

Reply:
89187f1 fixes this.